### PR TITLE
orcid: handle multiple ORCID putcodes for the same record

### PR DESCRIPTION
**Pull request by MJedr**
* INSPIR-3383

### CODE REVIEW

File: backend/inspirehep/orcid/domain_models.py [hunk 1]
```diff
@@ -204,16 +206,30 @@ def _post_or_put_work(self, putcode=None):
         response.raise_for_result()
         return response["putcode"]
 
+    def _delete_works_with_duplicated_putcodes(self, cached_putcodes_recids):
+        unique_putcodes_recids = {}
+        for fetched_putcode, fetched_recid in cached_putcodes_recids:
+            if fetched_recid in unique_putcodes_recids.values():
```
Comment (michamos) [inline backend/inspirehep/orcid/domain_models.py:212]: accessing a dictionary by keys is fast (O(1)), accessing a dictionary by values is slow (O(len(dict))). So it's better to invert the dict here so you can do
```suggestion
            if fetched_recid in unique_recids_putcodes:
```

File: backend/inspirehep/orcid/domain_models.py [hunk 2]
```diff
@@ -204,16 +206,30 @@ def _post_or_put_work(self, putcode=None):
...
+                self._delete_work(fetched_putcode)
+            else:
+                unique_putcodes_recids[fetched_putcode] = fetched_recid
+        return [(putcode, recid) for putcode, recid in unique_putcodes_recids.items()]
```
Comment (michamos) [inline backend/inspirehep/orcid/domain_models.py:216]: why not use the previously built dict as a cache instead?

File: backend/tests/integration/orcid/test_orcid_domain_models.py [hunk 3]
```diff
@@ -265,6 +264,20 @@ def test_push_new_work_already_existing_duplicated_external_identifier_exception
...
+    ):
+        with override_config(
+            ORCID_APP_CREDENTIALS={"consumer_key": "0000-0001-8607-8906"}
+        ):
+            self.add_work(
+                "test_orcid_domain_models_TestOrcidPusherPostNewWork_conflicting_no_recid.xml"
+            )
+            pusher = domain_models.OrcidPusher(self.orcid, self.recid, self.oauth_token)
+            pusher.push()
+            assert mock.called
```
Comment (michamos) [inline backend/tests/integration/orcid/test_orcid_domain_models.py:279]: why mock this? I would do it like the other tests and check the cassette to make sure the DELETE is sent and the response is ok, instead you should add 
```suggestion
            assert self.cache.has_work_content_changed(self.inspire_record)
```

Comment (MJedr) [reply backend/tests/integration/orcid/test_orcid_domain_models.py:279]: I am not sure if ```self.cache.has_work_content_changed(self.inspire_record)``` will be good there. We are randomly choosing the record which is going to be deleted and those records might change. It is why I preferred to check if one of the records was deleted.

Comment (michamos) [reply backend/tests/integration/orcid/test_orcid_domain_models.py:279]: at the end of the day, the push should succeed, so the record will have been pushed and the cache will agree with the current version.

Comment (MJedr) [reply backend/tests/integration/orcid/test_orcid_domain_models.py:279]: So it should be ```assert not self.cache.has_work_content_changed(self.inspire_record)``` as the record doesn't change.

Comment (michamos) [reply backend/tests/integration/orcid/test_orcid_domain_models.py]: indeed, I forgot the `not` :facepalm:

File: backend/tests/integration/orcid/cassettes/TestOrcidPusherPostNewWork.test_push_new_work_already_existing_and_delete_duplicated_records_different_putcodes.yaml [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+      - nginx/1.16.1
+      X-Content-Type-Options:
+      - nosniff
+      X-Frame-Options:
+      - DENY
+      X-XSS-Protection:
+      - 1; mode=block
+    status:
+      code: 409
+      message: Conflict
```
Comment (michamos) [inline backend/tests/integration/orcid/cassettes/TestOrcidPusherPostNewWork.test_push_new_work_already_existing_and_delete_duplicated_records_different_putcodes.yaml:118]: I had a second thought about this, but after thinking about it a bit more (and remembering tests are not isolated), it's fine :+1:

Comment (MJedr) [reply backend/tests/integration/orcid/cassettes/TestOrcidPusherPostNewWork.test_push_new_work_already_existing_and_delete_duplicated_records_different_putcodes.yaml:118]: Yes, the thing is that we are pushing the work which already exists (twice) in orcid, so to cache all author putcodes and delete duplicated record from orcid there must be 409 (WorkAlreadyExists) before.

Comment (MJedr) [reply backend/tests/integration/orcid/cassettes/TestOrcidPusherPostNewWork.test_push_new_work_already_existing_and_delete_duplicated_records_different_putcodes.yaml:118]: Yes, the thing is that we are pushing the work which already exists (twice) in orcid, so to cache all author putcodes and delete duplicated record from orcid there must be 409 (WorkAlreadyExists) before.

**Pull request merged**